SpringBoot 프로젝트 생성 & SQL쿼리 첨부

devNana222 opened the pull request:
프로젝트 생성했습니다.

kilhyeonjun committed:
docs: 프로젝트 브랜치 전략 및 작업 가이드라인 업데이트

- 브랜치 전략 상세 설명 추가
- 작업 흐름 및 코드 리뷰 프로세스 명확화
- 커밋 메시지 작성 지침 추가
- 주의사항 섹션 보완

devNana222 committed:
SpringBoot 프로젝트 생성 & SQL쿼리 첨부

Commit:
회원가입 API 생성

Commit:
gitignore, build.gradle

Commit:
.

kilhyeonjun commented on .gitignore:
```diff
@@ -0,0 +1,199 @@
+# Covers JetBrains IDEs: IntelliJ, RubyMine, PhpStorm, AppCode, PyCharm, CLion, Android Studio, WebStorm and Rider
+# Reference: https://intellij-support.jetbrains.com/hc/en-us/articles/206544839
+
+.gradle
+.idea
```
이게 적용이 안된 것 같아요.
https://yongdev91.tistory.com/11
해당 글 참고해서 git에서 제거 하면 좋을 것 같아요.

kilhyeonjun commented on src/main/.gitignore:
```diff
@@ -0,0 +1,37 @@
+HELP.md
```
root 에 gitignore가 있어서 불필요할 것 같아요.

kilhyeonjun commented on src/main/java/registration/entity/UserInfo.java:
```diff
@@ -0,0 +1,55 @@
+    private String regIp;
+
+
+    public String getUserName() {
+        return userName;
+    }
+
+    public void setUserName(String userName) {
+        this.userName = userName;
+    }
+
+    public String getPassword() {
+        return password;
+    }
+
+    public void setPassword(String password) {
+        this.password = password;
+    }
+
+    public String getRegId() {
+        return regId;
+    }
+
+    public void setRegId(String regId) {
+        this.regId = regId;
+    }
```
룸북의 `@Data` 어노테이션은 `@Getter` `@Setter` 등 여러 어노테이션이 합쳐져있어요.
`@Data` 사용을 지양하기 하나 사용하신다면 작성해주신 getter setter 메소드가 불필요해 보입니다.

kilhyeonjun commented on src/main/java/registration/userService/GetUserInfo.java:
```diff
@@ -0,0 +1,60 @@
+
+    //    UserInfo user = new UserInfo();
+
+        if(userRepository.findOneByUserName(user.getUserName()) != null)
```
username이 일반 로그인시에 id값이기에 중복을 불가능하게 설정해주신 것 같은데 Table에 해당 컬럼을 유니크로 설정해서 동시성 문제로 인해 중복값이 들어가는 것을 방지해주시면 좋을 것 같아요.

kilhyeonjun commented on src/main/java/registration/userService/GetUserInfo.java:
```diff
@@ -0,0 +1,60 @@
+//            return false;
+        DateFormat now = new SimpleDateFormat("YYYY-MM-dd HH:mm:ss");
+
+        user.setUserName(user.getUserName());
+        user.setPassword(user.getPassword());
+        user.setRegId("registerUser");
+        user.setRegDate(now.format(new Date()));
+
+        user.setRegIp("");
```
이미 클라이언트에 받아온 값을 다시 할당해주는 것 같은데 불필요한 코드인 것 같아요.

kilhyeonjun commented on src/main/java/registration/controller/UserController.java:
```diff
@@ -0,0 +1,27 @@
+
+    @PostMapping("/register")
+    public ResponseEntity<String> registerUser(@RequestBody UserInfo user){
+        if (userInfo.registerUser(user)) {
+            return ResponseEntity.ok("User registered successfully!");
+        } else {
+            return ResponseEntity.badRequest().body("Username "+user.getUserName()+" already exists!");
+        }
```
성공 실패여부는 컨트롤러에서보단 서비스 레이어해서 해야할 작업인 것 같아요.
서비스 레이어에 return을 굳이 boolean으로 하지 말고 error throw하는 형태로 가도 좋을 것 같아요.

kilhyeonjun commented:
여기까지 확인마쳤습니다. 위에 코멘트의 반영할지말지 여부는 진행 상황에 따라 해주시고 참고만 해주시면 좋을 것 같습니다.

Commit:
게시판 등록 API(BoardList)생성, PR 피드백 반영

Commit:
gitignore 적용

kilhyeonjun commented on gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -0,0 +1,7 @@
+distributionBase=GRADLE_USER_HOME
```
.gradle

kilhyeonjun commented on board/README.md:
```diff
@@ -0,0 +1,33 @@
+# 게시판 프로젝트
```
root에 있는 readme랑 겹치는 것 같은데 이건 제거하셔도 될 것 같아요.

kilhyeonjun commented on src/main/java/hh99/BoardProject/registration/controller/BoardController.java:
```diff
@@ -0,0 +1,34 @@
+
+
+@RestController
+@RequestMapping("/api/board")
```
Restful 한 API 설계를 위해선 자원의 경우 단수가 아닌 복수로 사용해주시는 게 좋습니다. 
`/api/board` -> `/api/boards`

kilhyeonjun commented on src/main/java/hh99/BoardProject/registration/controller/BoardController.java:
```diff
@@ -0,0 +1,34 @@
+    @Autowired
+    private BoardService boardService;
+
+    @PostMapping("/register")
```
Restful 한 API에선 자원의 행위를 http 메소드로 구분합니다.
Post의 경우엔 그 자원을 생성하는 행위라 `/register`는 없어도 되지않을까 싶습니다.

![image](https://github.com/user-attachments/assets/aa698229-aa86-447d-899b-5a0b373afe40)

kilhyeonjun commented on src/main/java/hh99/BoardProject/registration/controller/UserController.java:
```diff
@@ -0,0 +1,28 @@
+    @Autowired
+    private UserService userService;
+
+    @PostMapping("/register")
```
^2

kilhyeonjun commented on src/main/java/hh99/BoardProject/registration/entity/BoardList.java:
```diff
@@ -0,0 +1,51 @@
+@Table(name="BOARD_LIST")
+public class BoardList {
+
+    @Id
+    @Column(name="title", nullable = false)
+    private String title;
```
음 title이 Id(PK)이면 중복이 안되서 별도의 Id값을 빼주셔야할 것 같습니다.

kilhyeonjun commented on src/main/java/hh99/BoardProject/registration/entity/UserInfo.java:
```diff
@@ -0,0 +1,31 @@
+@Table(name="USER_INFO")
+public class UserInfo {
+
+    @Id
+    @Column(name="user_name", nullable = false, unique = true)
+    private String userName;
```
^2

kilhyeonjun commented on src/main/java/hh99/BoardProject/registration/userService/BoardService.java:
```diff
@@ -0,0 +1,40 @@
+            list.setUseYn("Y");
+
+            boardListRepository.save(list);
+            return 0;
```
객체 생성 메소드의 경우엔 성공 여부를 반환하기 보단 보통 객체의 id값이나 생성된 객체를 반환합니다.
<img width="371" alt="image" src="https://github.com/user-attachments/assets/72b58d2a-1bd5-4c55-a2ab-c0a22d0d50c5">

하지만 과제물 요구사항에선 객체의 반환을 요구하게 맞춰서 해주시면 좋을 것 같아요. 
나중에 제출하는 것 같더라구여

https://www.notion.so/teamsparta/Spring-584d17598621430ead3978930a842f1f

kilhyeonjun commented on src/main/java/hh99/BoardProject/registration/userService/BoardService.java:
```diff
@@ -0,0 +1,40 @@
+    BoardListRepository boardListRepository;
+    UserInfoRepository userInfoRepository;
+
+    public Integer registPost(BoardList list)
```
entity 네이밍이 BoardList라서 param을 list로 명시해주신 것 같은데 board와 같이 단수 네이밍으로 가야 좋을 것 같아요.
list로만 봤을 때 대량의 게시글을 생성할 것 처럼 혼란을 줄 수 있을 것 같습니다.

kilhyeonjun commented on src/main/java/hh99/BoardProject/registration/userService/UserService.java:
```diff
@@ -0,0 +1,37 @@
+        user.setRegIp("");
+
+        userRepository.save(user);
+        return true; //회원가입 성공
```
^2

kilhyeonjun commented on src/main/java/hh99/BoardProject/registration/userService/BoardService.java:
```diff
@@ -0,0 +1,40 @@
+@Service
+public class BoardService {
+
+    @Autowired
```
최근에 autowired 보단 constructor injection 방식을 선호하고 있는데 이 부분 찾아보면 도움이 되실 것 같습니다.

kilhyeonjun commented on src/main/java/hh99/BoardProject/registration/userService/BoardService.java:
```diff
@@ -0,0 +1,40 @@
+public class BoardService {
+
+    @Autowired
+    BoardListRepository boardListRepository;
+    UserInfoRepository userInfoRepository;
```
외부에서 해당 객체를 참조할 수 없게 private를 명시해주세요!

kilhyeonjun reviewed (commented):
아래의 코멘트 한번 확인해주시고

https://www.notion.so/teamsparta/Spring-584d17598621430ead3978930a842f1f

해당 과제의 필수 요구사항 확인해주시고 맞춰서 작성해주시면 좋을 것 같습니다.
완료 후 배포하고 제출하는 과정도 있었던 것 같아요.

Commit:
전체 게시글 목록 조회 API 생성 & 피드백 반영

1. PostMapping 삭제하였습니다.
2. @Autowired 주입 대신 생성자 주입 방식을 사용하였습니다. (@RequiredArgsConstructor를 사용하면 생성자를 자동으로 생성해준다고합니다.)
3. 게시판 등록에 userName대신 regId가 좋을 것 같아 변경하였습니다.
4. GetterSetter을 제거하고 @Data 어노테이션으로 대체하였습니다.

5. 전체 게시글 조회하는 API 생성하였습니다.
5-1. 단순 findAll로 전체 조회하는 getAllPosts()
5-2. Native Query : Entity에 DTO를 생성하여 @Query 로 select하여 생성하는 방법 > 그치만 이건 API json 반환 시 원하는 컬럼만 반환해주지는 않았습니다. 때문에 사용되지 않는 방법이라고 합니다.
5-3. Projection Interface를 이용해서 하는 방법. > 조회하는 테이블이 ROOT테이블 일 경우 전체 컬럼을 조회하는 것이 아닌, 원하는 컬럼만을 뽑아낼 수도 있기에 성능개선측면과 응답성 측면에서 장점이 있습니다.

https://learngoeson.tistory.com/51

kilhyeonjun commented on src/main/java/hh99/BoardProject/registration/userService/BoardService.java:
```diff
@@ -0,0 +1,58 @@
+
+
+
+    private final BoardListRepository boardListRepository;
+
+    public BoardService(BoardListRepository boardListRepository) {
+        this.boardListRepository = boardListRepository;
+    }
```
lombok 사용중이 신 것 같은데 @RequiredArgsConstructor 어노테이션 사용해주셔도 좋을 것 같아요!

kilhyeonjun replied:
lombok 사용 중이신 것 같은데 @RequiredArgsConstructor 어노테이션활용하면 생성자 구문을 생략할 수 있습니다!

kilhyeonjun replied:
lombok 사용 중이신 것 같은데 @RequiredArgsConstructor 어노테이션활용하면 생성자 구문을 생략할 수 있습니다!

kilhyeonjun commented on src/main/java/hh99/BoardProject/registration/userService/UserService.java:
```diff
@@ -0,0 +1,40 @@
+public class UserService {
+
+
+    private final UserInfoRepository userRepository;
+
+    public UserService(UserInfoRepository userRepository) {
+        this.userRepository = userRepository;
+    }
```
^2

kilhyeonjun commented on src/main/java/hh99/BoardProject/registration/repository/BoardListRepository.java:
```diff
@@ -0,0 +1,21 @@
+import java.util.List;
+
+@Repository
+public interface BoardListRepository extends JpaRepository<BoardList, String> {
```
JpaRepository의 제레닉의 첫번째 인값은 다룰 entity, 두번째 인자값은 pk type입니다.
BoardList의 pk가 Integer type 이다 보니 변경이 필요할 것 같아요.

kilhyeonjun commented on src/main/java/hh99/BoardProject/registration/repository/UserInfoRepository.java:
```diff
@@ -0,0 +1,11 @@
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface UserInfoRepository extends JpaRepository<UserInfo, String> {
```
^2

kilhyeonjun commented on src/main/java/hh99/BoardProject/registration/userService/BoardService.java:
```diff
@@ -0,0 +1,58 @@
+        //    //존재하지 않는 유저
+       //     return -10;
+
+        if(post.getPassword() == null || post.getPassword().isEmpty()){
```
password와 같은 민감한 데이터는 암호화된 값을 저장합니다.
암호화 방식엔 단방향 양방향이 있는데 password는 단방향 패턴을 사용하고 있는데 이 부분도 찾아보시면 도움이 될 거에요!
사전 스터디다보니 password 암호화 적용 여부는 예진님꼐서 편하시게 정하시고 가시면 될 것 같습니다. (굳이 적용안해도되요!)

kilhyeonjun reviewed (approved):
좋습니당
